master/pkg/schemas/expconf/log_pattern_config.go

```diff
@@ -4,86 +4,278 @@ import (
+// can be associated with a pattern. For example:
+//
+//   - pattern: a
+//     action: cancel_retries
+//   - pattern: a
+//     action: exclude_node
+//
+// An actions field is available now. Multiple entires can be associated with a pattern. No more
+// duplicated pattern in log policies config. For example:
+//
+//   - pattern: a
+//     actions:
+//   - cancel_retries
+//   - exclude_node
+func (b *LogPoliciesConfigV0) UnmarshalJSON(data []byte) error {
+	type DefaultParser LogPoliciesConfigV0
+	var jsonItems DefaultParser
+	if err := json.Unmarshal(data, &jsonItems); err != nil {
+		return errors.Wrapf(err, "failed to parse runtime items")
+	}
+
+	// Merge LogPolicyV0s with the same pattern into one
+	patternToActions := make(map[string]set.Set[LogActionV0])
+	for _, jsonItem := range jsonItems {
+		if jsonItem.RawLegacyAction != nil {
+			return fmt.Errorf("legacy action field expected to be nil: %+v", jsonItem)
+		}
+
+		if _, ok := patternToActions[jsonItem.RawPattern]; !ok {
+			patternToActions[jsonItem.RawPattern] = set.New[LogActionV0]()
+		}
+
+		for _, a := range jsonItem.RawActions {
+			actions := patternToActions[jsonItem.RawPattern]
+			actions.Insert(a)
+		}
+	}
+
+	for p, actions := range patternToActions {
+		lp := LogPolicyV0{
+			RawPattern: p,
+			RawActions: actions.ToSlice(),
+		}
+		*b = append(*b, lp)
+	}
+
+	return nil
+}
+
 // LogPolicyV0 is an action to take if we match against trial logs.
 //
 //go:generate ../gen.sh
 type LogPolicyV0 struct {
-	RawPattern string `json:"pattern"`
+	RawPattern      string             `json:"pattern"`
+	RawLegacyAction *LogLegacyActionV0 `json:"action,omitempty"`
+	RawActions      []LogActionV0      `json:"actions,omitempty"`
+}
 
-	RawAction LogActionV0 `json:"action"`
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (b *LogPolicyV0) UnmarshalJSON(data []byte) error {
+	type DefaultParser LogPolicyV0
+	var jsonItems DefaultParser
+	if err := json.Unmarshal(data, &jsonItems); err != nil {
+		return fmt.Errorf("failed to parse log policy: %w", err)
+	}
+	b.RawPattern = jsonItems.RawPattern
+
+	// Backward compatiblity. Move data in the legacy action field to the newer actions field.
+	if jsonItems.RawLegacyAction != nil && jsonItems.RawActions == nil {
+		if jsonItems.RawLegacyAction.RawCancelRetries != nil {
+			b.RawActions = []LogActionV0{{CancelRetries: &LogActionCancelRetries{}}}
+		} else if jsonItems.RawLegacyAction.RawExcludeNode != nil {
+			b.RawActions = []LogActionV0{{ExcludeNode: &LogActionExcludeNode{}}}
+		} else {
+			return fmt.Errorf("invalid legacy log action: %+v", jsonItems.RawLegacyAction)
+		}
+		b.RawLegacyAction = nil
+	} else if jsonItems.RawLegacyAction == nil && jsonItems.RawActions != nil {
+		b.RawActions = jsonItems.RawActions
+	} else {
+		return fmt.Errorf("invalid log policy: %+v", jsonItems)
+	}
+
+	return nil
 }
 
 // LogActionV0 is a policy to take after matching.
 //
 //go:generate ../gen.sh
+type LogLegacyActionV0 struct {
```

Comment:
I think you shouldn't need this type in the go code.  I _think_ you just need to allow it in the json schema, and you need to do the shimming in the new LogPolicyV0 unmarshaling.  But I don't see why you actually need this object in go, ever.